internal/resolution/client/npm_registry_client.go

```diff
@@ -0,0 +1,267 @@
+
+	peerType := dep.NewType()
+	peerType.AddAttr(dep.Scope, "peer")
+	addDeps(dependencies.PeerDependencies, peerType)
+
+	// The resolver expects bundleDependencies to be present as regular
+	// dependencies with a "*" version specifier, even if they were already
+	// in the regular dependencies.
+	bundleType := dep.NewType()
+	bundleType.AddAttr(dep.Scope, "bundle")
+	for _, name := range dependencies.BundleDependencies {
+		deps = append(deps, resolve.RequirementVersion{
+			Type: bundleType,
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.NPM,
+					Name:   name,
+				},
+				VersionType: resolve.Requirement,
+				Version:     "*",
+			},
+		})
+	}
+
+	// Correctly resolving the bundled dependencies would require downloading the package
+	// call the fallback deps.dev client to get the bundled dependencies with mangled names
+	if len(dependencies.BundleDependencies) > 0 {
+		fallbackReqs, err := c.fallback.Requirements(ctx, vk)
+		if err != nil {
+			// TODO: make some placeholder if the package doesn't exist in the deps.dev data
+			return nil, err
+		}
+		for _, req := range fallbackReqs {
+			if strings.Contains(req.Name, ">") {
+				deps = append(deps, req)
+			}
+		}
+	}
+
+	resolve.SortDependencies(deps)
+
+	return deps, nil
+}
+
+func (c *NpmRegistryClient) MatchingVersions(ctx context.Context, vk resolve.VersionKey) ([]resolve.Version, error) {
+	if strings.Contains(vk.Name, ">") { // bundled dependencies, fallback to deps.dev client
+		return c.fallback.MatchingVersions(ctx, vk)
+	}
+
+	versions, err := c.api.Versions(ctx, vk.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if concVer, ok := versions.Tags[vk.Version]; ok {
+		// matched a tag, return just the concrete version of the tag
+		return []resolve.Version{{
+			VersionKey: resolve.VersionKey{
+				PackageKey:  vk.PackageKey,
+				Version:     concVer,
+				VersionType: resolve.Concrete,
+			},
+		},
+		}, nil
+	}
+
+	resVersions := make([]resolve.Version, len(versions.Versions))
+	for i, v := range versions.Versions {
+		resVersions[i] = resolve.Version{
+			VersionKey: resolve.VersionKey{
+				PackageKey:  vk.PackageKey,
+				Version:     v,
+				VersionType: resolve.Concrete,
+			},
+		}
+	}
+
+	return resolve.MatchRequirement(vk, resVersions), nil
+}
+
+func (c *NpmRegistryClient) PreFetch(ctx context.Context, imports []resolve.RequirementVersion, manifestPath string) {
+	// It doesn't matter if loading the cache fails
+	_ = c.LoadCache(manifestPath)
+
+	// Use the deps.dev client to fetch complete dependency graphs of our direct imports
+	for _, im := range imports {
+		// Get the preferred version of the import requirement
+		vks, err := c.MatchingVersions(ctx, im.VersionKey)
+		if err != nil || len(vks) == 0 {
+			continue
+		}
+
+		vk := vks[len(vks)-1]
+
+		// Make a request for the precomputed dependency tree
+		// TODO: avoid relying on DepsDevClient internals
```

Comment:
is this TODO done?

Reply:
Just changed the code a bit - now it's done.